DevCycle.SDK.Server.Common/Model/DevCycleUser.cs

```diff
@@ -336,48 +345,93 @@ public bool Equals(DevCycleUser input)
+
+            foreach (var (k, v) in context.AsDictionary().Select(x => (x.Key, x.Value)))
+            {
+                if (k.Equals("user_id") || k.Equals("targetingKey")) continue;
+
+                switch (k)
+                {
+                    case "email":
+                    case "name":
+                    case "language":
+                    case "country":
+                    case "appVersion":
+                        if (!v.IsString)
+                            continue;
+                        break;
+                    case "appBuild":
+                        if (!v.IsNumber)
+                            continue;
+                        break;
+                    case "customData":
+                    case "privateCustomData":
+                        if (v.IsStructure)
+                        {
+                            var str = v.AsStructure;
+                            foreach (var (cdK, cdV) in str.AsDictionary().Select(x => (x.Key, x.Value)))
+                            {
+                                if (!cdV.IsString && !cdV.IsNumber && !cdV.IsBoolean && !cdV.IsNull)
+                                    throw new System.Exception(
+                                        "DevCycleUser only supports flat customData properties of type string / number / boolean / null");
+
+                                switch (k)
+                                {
+                                    case "privateCustomData":
+                                        user.PrivateCustomData.Add(cdK, cdV.AsObject);
+                                        break;
+                                    case "customData":
+                                        user.CustomData.Add(cdK, cdV.AsObject);
+                                        break;
+                                }
+                            }
+
+                            continue;
+                        }
+
+                        break;
+                    default:
+                        // add to customdata
+                        if (!v.IsString && !v.IsNumber && !v.IsBoolean && !v.IsNull)
+                            throw new System.Exception(
+                                $"DevCycleUser only supports flat customData properties of type string / number / boolean / null. Attempted to submit {v}");
+                        user.CustomData[k] = v.AsObject;
+                        continue;
+                }
+
+                if (k.Equals("customData") || k.Equals("privateCustomData")) continue;
+
+                var propertyName = Char.ToUpperInvariant(k[0]) + k.Substring(1);
+                var propertyInfo = user.GetType().GetProperty(propertyName);
+                if (propertyInfo != null)
+                    propertyInfo.SetValue(user, v.AsObject, null);
+                else
+                {
+                    throw new System.Exception($"invalid property key referenced {propertyName} {v}");
```

Comment:
This probably should be a different exception

Reply:
Yeah, i'd throw something more specific would be nice.